### PR TITLE
Infinite retry strategy

### LINKED ISSUE #78: Provide an option to retry forever on 503 and 504 responses
The vast majority of exceptions that prawcore raises to my bot are 503 and 504 status codes when Reddit is experiencing outages. In this case, the only sensible thing to do is to wait a while and try again.

I think it would make sense for prawcore to provide this behavior as an option, rather than re-implementing it in every client that wants this behavior.

It probably should be optional, so that (e.g.) applications with user interfaces can just report the outage instead of stalling indefinitely.

Does it sound right for prawcore to do this?

### PR DESCRIPTION
**Pull request by leviroth**
Fixes #78.

The retry logic itself is in the first commit. I'm not really sure what the proper way is to expose this to PRAW, however, so the second commit will probably need some work.

Testing is also TODO.

### CODE REVIEW

**Comment by bboe** [comment]
First commit looks good. But this got me thinking, wasn't the original idea to conditionally retry forever on certain 5XX responses? In that case, does it even make sense to split out the general retry logic?

**Comment by leviroth** [comment]
>But this got me thinking, wasn't the original idea to conditionally retry forever on certain 5XX responses?

I think it makes sense to only do this for some of the responses. However, I don't think I'm very confident about which ones those should be.

File: prawcore/sessions.py [hunk 1]
```diff
@@ -78,6 +78,33 @@ def should_retry_on_failure(self):
...
+        return self._base_delay + 2 * random.random()
+
+    def __init__(self, base_delay=2, max_delay=60):
+        """Initialize the strategy.
+
+        :param base_delay: Approximate time to sleep on the next retry.
+        :param max_delay: Maximum (approximate) time to sleep on any retry.
+
+        """
+        self._base_delay = max(base_delay, max_delay)
```
Comment (christianhans) [inline prawcore/sessions.py]: This should be `self._base_delay = min(base_delay, max_delay)`

**Comment by christianhans** [comment]
I've found a small bug in this PR: https://github.com/praw-dev/prawcore/pull/81/files#r344321585

**Comment by leviroth** [comment]
Thanks! Fwiw, I got an email notification when you commented on the diff, without having to comment on the PR itself.

**Comment by christianhans** [comment]
Good to know, thanks!

File: prawcore/sessions.py (hunk 1 above)
Comment (PythonCoderAS) [reply prawcore/sessions.py:94]:
```suggestion
        self._base_delay = min(base_delay, max_delay)
```

Comment (leviroth) [reply prawcore/sessions.py:94]: It's amazing how consistently I screw that up.

**Comment by bboe** [comment]
Just wanted to check in on the status of this PR. Is it worth keeping open?

**Comment by leviroth** [comment]
Realistically, I don't think I'm going to resume work on it.

**Pull request closed without merging**